### ISSUE TITLE
Inference with PRIMERA much slower than inference with PRIMERA-multinews

### ISSUE THREAD

**Issue by JohnGiorgi** [allenai/PRIMER#11]
Hi!

I have noticed something strange. Inference with `allenai/PRIMERA` is much slower (as much as 5X!) than inference with `allenai/PRIMERA-multinews`. I have a [notebook benchmarking this effect here](https://colab.research.google.com/drive/18bssWN9H3gdxK7cQSlyTvZsF2wcvpZ6e?usp=sharing). I checked that their model configs are identical.

The problem is most likely related to the decoder/generation because it only occurs with the `max_length` argument to `model.generate()` is large (say `1024`) and not when it is small (say `64`). Here are some benchmarks using some random examples from MultiNews:

_With a batch size of 4 and max length of inputs 1024:_

__Max length of outputs: 64__
- `PRIMERA` CUDA time total: 758.402ms
- `PRIMERA-multinews` CUDA time total: 753.125ms
- Slowdown: ~0X

__Max length of outputs: 512__
- `PRIMERA` CUDA time total: 3.682s
- `PRIMERA-multinews` CUDA time total: 1.572s
- Slowdown: ~2X

__Max length of outputs: 1024__
- `PRIMERA` CUDA time total: 7.676s
- `PRIMERA-multinews` CUDA time total: 1.542s
- Slowdown: ~5X

Do you have any idea what might be causing this?

---

__EDIT__: While I don't know why this is _partially_ explained by the fact that `allenai/PRIMERA` does not appear to use the `global_attention_mask`. Running the model with and without providing `global_attention_mask` leads to the same inference times, while with `allenai/PRIMERA-multinews`, providing `global_attention_mask` leads to a 30% reduction in inference time. However, `allenai/PRIMERA-multinews` without a `global_attention_mask` is still almost 3X faster than `allenai/PRIMERA`, so this couldn't entirely explain the difference.

**Comment by Wendy-Xiao** [comment]
Hi, 

I don't have any idea about the global mask problem. Are you using the same scripts for data loading and training/inference for both models? I think the difference between the two models is only the model parameters, and I'm not sure why `allenai/PRIMERA` does not use `global_attention_mask`... As the global masks are applied on the `<s>``</s>` `<doc-sep>` tokens for all of PRIMERA models (both pre-trained and fine-tuned), if you are using the same input for both models, then both of them should have the same `global_attention_mask`.

About the time, did you compare the avg length of generated summaries? One possible reason I can think of, the `PRIMERA` model tend to generate longer summaries, which takes a longer time, and `PRIMERA-multinews` tend to generate shorter summaries (should be similar to the avg length of the summaries in the multi-news dataset).

**Comment by JohnGiorgi** [comment]
Closing this, thanks @Wendy-Xiao! I think it can, in fact be explained by the longer generation on average from PRIMERA vs PRIMERA-multinews.
